hack/remove-members.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+var dryrun bool
+var repoRoot string
+
+func main() {
+	flag.StringVar(&repoRoot, "root", ".", "Root of the repository")
+	flag.BoolVar(&dryrun, "dryrun", true, "Enable Dryrun or not")
+
+	flag.Parse()
+
+	if len(flag.Args()) < 1 {
+		fmt.Print("Usage: remove-members [flags] <file-containing-members-list>\n\nFlags:\n\t-root\tstring\tpath to root directory of the repository\n\t-dryrun\tboolean\tEnable/Disable dryrun (Default: Enabled)")
+		os.Exit(0)
+	}
+
+	memberList, err := readMemberList(flag.Args()[0])
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	configPath := repoRoot + "/config"
+
+	err = removeMembers(memberList, configPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+//readMemberList reads the list of members to be removed from the given filepath
+func readMemberList(path string) ([]string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var members []string
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		members = append(members, scanner.Text())
+	}
+	return members, scanner.Err()
+}
+
+//removeMembers walks through the config directory and removes the occurences of the given member name
+func removeMembers(memberList []string, configPath string) error {
+	for _, member := range memberList {
+		var orgs, teams []string
+		count := 0
+		fmt.Print(member)
+
+		err := filepath.Walk(configPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+
+			if matched, err := filepath.Match("*.yaml", filepath.Base(path)); err != nil {
+				return err
+			} else if matched {
+				removed, err := removeMemberFromFile(member, path)
+				if err != nil {
+					return err
+				}
+
+				//Record the org/team name when a member is removed from it
+				if removed {
+					count++
+					if info.Name() == "org.yaml" {
+						orgs = append(orgs, filepath.Base(filepath.Dir(path)))
+					}
+					if info.Name() == "teams.yaml" {
+						teams = append(teams, filepath.Base(filepath.Dir(path)))
+					}
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		sort.Strings(orgs)
+		sort.Strings(teams)
+
+		fmt.Printf("\n Orgs: %v\n Teams: %v\n Number of occurences: %d\n", orgs, teams, count)
+
+		//Proceed to committing changes if member is actually removed from somewhere
+		if count > 0 {
+			commitRemovedMembers(member, orgs, teams)
+		}
+	}
+
+	return nil
+}
+
+func removeMemberFromFile(member string, path string) (bool, error) {
+
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	re := regexp.MustCompile(`(\s+)?- ` + member + `(.*)?`)
+
+	if re.Match(content) {
```

Comment:
we can probably count how many times `member` occurs in the `path` and return the count. This will ensure that `count` is accurate.

Reply:
Do you mean to count the number of matches in the said file?

Reply:
> Do you mean to count the number of matches in the said file?

Yeah :)

Reply:
Not a blocker for this, but potentially something in the future. We can marshal/unmarshal yaml while leaving the comments (https://github.com/go-yaml/yaml/issues/132). This would also let us get team names directly.

Reply:
@mrbobbytables -- I had a dig at this earlier by parsing the org files using [k8s.io/test-infra/prow/config/org.Config](https://github.com/kubernetes/test-infra/blob/master/prow/config/org/org.go), but there seemed to be something off there.

The processed org config when written back to disk had a lot of diff compared to the original file.

We can possibly dig deeper into it later.

Reply:
It's been a bit since I looked at it , but potentially whitespace =/ since yaml supports some fuzzy inclusion:
```
foo:
  - bar
```
is the same as:
```
foo:
- bar
```

Theres what people have written (both formats) and what go will want to format it as.

Reply:
That as well. we may want to bite the bullet with a big initial diff as the subsequent updates _should_ be deterministic .

One other thing was, in order to process comments, the structs in `k8s.io/test-infra/prow/config/org.Config` should have fields with `yaml.Node` as field type and then have post-processing helper method. I had taken a step back because of that complexity.

Reply:
Yeah, what I'd probably do at that point is get away from editing by hand, and then have an add-member function or have a make command that ensures proper formatting before commit. Out of scope for this one though^^;;;

Reply:
Adding to whitespaces and comments related changes, The substantial amount of diff is also created due to the difference in order of the yaml properties.
I was trying to parse the yaml files and removing members in a more structured way, I also faced this.

```diff
diff --git a/config/kubernetes-incubator/org.yaml b/config/kubernetes-incubator/org.yaml
index baf4ca76..e2da8a87 100644
--- a/config/kubernetes-incubator/org.yaml
+++ b/config/kubernetes-incubator/org.yaml
@@ -1,10 +1,3 @@
-name: Kubernetes Incubator
-description: Kubernetes Incubator
-default_repository_permission: read
-has_organization_projects: true
-has_repository_projects: true
-members_can_create_repositories: false
-billing_email: github@kubernetes.io
 admins:
 - cblecker
 - fejta
@@ -15,3 +8,10 @@ admins:
 - nikhita
 - spiffxp
 - thelinuxfoundation
+billing_email: github@kubernetes.io
+default_repository_permission: read
+description: Kubernetes Incubator
+has_organization_projects: true
+has_repository_projects: true
+members_can_create_repositories: false
+name: Kubernetes Incubator
```
It seems to be sorting the fields.

I currently implemented this using `kubernetes-sigs/yaml`.